Modules/_ctypes/stgdict.c

```diff
@@ -356,32 +357,63 @@ PyCStructUnionType_update_stgdict(PyObject *type, PyObject *fields, int isStruct
     if (fields == NULL)
         return 0;
 
-#ifdef WORDS_BIGENDIAN
-    big_endian = PyObject_HasAttrString(type, "_swappedbytes_") ? 0 : 1;
-#else
-    big_endian = PyObject_HasAttrString(type, "_swappedbytes_") ? 1 : 0;
-#endif
+    tmp = PyObject_GetAttrString(type, "_swappedbytes_");
+    if (tmp) {
+        Py_DECREF(tmp);
+        big_endian = !PY_BIG_ENDIAN;
+    }
+    else if (PyErr_ExceptionMatches(PyExc_AttributeError)) {
+        PyErr_Clear();
+        big_endian = PY_BIG_ENDIAN;
+    }
+    else {
+        return -1;
+    }
 
-    use_broken_old_ctypes_semantics = \
-        PyObject_HasAttrString(type, "_use_broken_old_ctypes_structure_semantics_");
+    tmp = PyObject_GetAttrString(type, "_use_broken_old_ctypes_structure_semantics_");
+    if (tmp) {
+        Py_DECREF(tmp);
+        use_broken_old_ctypes_semantics = 1;
+    }
+    else if (PyErr_ExceptionMatches(PyExc_AttributeError)) {
+        PyErr_Clear();
+        use_broken_old_ctypes_semantics = 0;
+    }
+    else {
+        return -1;
+    }
 
-    isPacked = PyObject_GetAttrString(type, "_pack_");
-    if (isPacked) {
-        pack = _PyLong_AsInt(isPacked);
-        if (pack < 0 || PyErr_Occurred()) {
-            Py_XDECREF(isPacked);
-            PyErr_SetString(PyExc_ValueError,
-                            "_pack_ must be a non-negative integer");
+    tmp = PyObject_GetAttrString(type, "_pack_");
+    if (tmp) {
+        isPacked = 1;
+        pack = _PyLong_AsInt(tmp);
+        Py_DECREF(tmp);
+        if (pack < 0) {
+            if (!PyErr_Occurred() ||
+                PyErr_ExceptionMatches(PyExc_TypeError) ||
+                PyErr_ExceptionMatches(PyExc_OverflowError))
+            {
+                PyErr_SetString(PyExc_ValueError,
+                                "_pack_ must be a non-negative integer");
+            }
             return -1;
         }
-        Py_DECREF(isPacked);
-    } else
+    }
+    else if (PyErr_ExceptionMatches(PyExc_AttributeError)) {
         PyErr_Clear();
+        isPacked = 0;
+        pack = 0;
+    }
+    else {
+        return -1;
+    }
 
-    len = PySequence_Length(fields);
+    len = PySequence_Size(fields);
```

Comment:
Why `Length` &rarr; `Size`?

Reply:
PySequence_Size is more commonly used. PySequence_Length is just an outdated alias of PySequence_Size.